konnectivity-socks5-proxy/main.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+
+	socks5 "github.com/armon/go-socks5"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/types"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/apiserver-network-proxy/pkg/util"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func main() {
+	cmd := &cobra.Command{
+		Use: "konnectivity-socks5-proxy",
+		Run: func(cmd *cobra.Command, args []string) {
+			cmd.Help()
+			os.Exit(1)
+		},
+	}
+	cmd.AddCommand(NewStartCommand())
+
+	if err := cmd.Execute(); err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+}
+
+func NewStartCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "run",
+		Short: "Runs the konnectivity socks5 proxy server.",
+		Long: ` Runs the konnectivity socks5 proxy server.
+		This proxy accepts request and tunnels them through the designated Konnectivity Server.
+		When resolving hostnames, the proxy will attempt to derrive the Cluster IP Address from
+		a Kubernetes Service using the provided KubeConfig. If the IP address
+		cannot be resolved from a service, the system DNS is used to resolve hostnames.
+		`,
+	}
+
+	var proxyHostname string
+	var proxyPort int
+	var servingPort int
+	var caCertPath string
+	var clientCertPath string
+	var clientKeyPath string
+
+	cmd.Flags().StringVar(&proxyHostname, "konnectivity-hostname", "konnectivity-server-local", "The hostname of the konnectivity service.")
+	cmd.Flags().IntVar(&proxyPort, "konnectivity-port", 8090, "The konnectivity port that socks5 proxy should connect to.")
+	cmd.Flags().IntVar(&servingPort, "serving-port", 8090, "The port that socks5 proxy should serve on.")
+
+	cmd.Flags().StringVar(&caCertPath, "ca-cert-path", "/etc/konnectivity-proxy-tls/ca.crt", "The path to the konnectivity client's ca-cert.")
+	cmd.Flags().StringVar(&clientCertPath, "tls-cert-path", "/etc/konnectivity-proxy-tls/tls.crt", "The path to the konnectivity client's tls certificate.")
+	cmd.Flags().StringVar(&clientKeyPath, "tls-key-path", "/etc/konnectivity-proxy-tls/tls.key", "The path to the konnectivity client's private key.")
+
+	cmd.Run = func(cmd *cobra.Command, args []string) {
+		fmt.Println("Starting proxy...")
+
+		client, err := client.New(ctrl.GetConfigOrDie(), client.Options{})
```

Comment:
This will be the wrong client as this container will run on the management side. You need a client to the guest cluster. You'll need to pass a kubeconfig filename as an argument and build the client using that.

Reply:
I don't believe that this is true as I've replicated how the OLM Deployment uses the guest cluster kubeconfig, process described below:

1. The `KUBECONFIG` Environment Variable is set to `/etc/openshift/kubeconfig/kubeconfig` here: https://github.com/openshift/hypershift/pull/483/files#diff-2df48c14467c38423809becf351b2491579fa74eb5b36202ababb386a94f737aR44-R45
2. The `Service Network Admin` secret is added as a volume to the deployment here: https://github.com/openshift/hypershift/pull/483/files#diff-2df48c14467c38423809becf351b2491579fa74eb5b36202ababb386a94f737aR34-R36
3. The volume is then mounted to the pod at the location described in step 1 here: https://github.com/openshift/hypershift/pull/483/files#diff-2df48c14467c38423809becf351b2491579fa74eb5b36202ababb386a94f737aR52-R54

I have tested this change by:
1. Hardcoding the images in the olm/catalog deployment templates and building the binaries
2. Scaling the HyperShift Operator and the Control Plane Operator to 0
3. Running the Control Plane Operator
4. Installing a CatalogSource on the guest cluster and confirming that the hostname is resolved to the correct IP Address

Reply:
One last note, the [GetConfigOrDie function](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client/config#GetConfigOrDie) will retrieve the config from the KUBECONFIG variable because a flag is not set.

Reply:
ahh ic, got it. Had not noticed the KUBECONFIG in the environment. Thx!